### PR TITLE
Upgrade to Kubernetes 1.27 and Go 1.20

### PR DESCRIPTION
**Pull request by nrb**
**What this PR does / why we need it**:
Go 1.20 was required to satisfy constraints for the following modules:

* k8s.io/apimachinery/pkg/util/managedfields/internal
* k8s.io/apiserver/pkg/authentication/token/cache
* k8s.io/apiserver/pkg/server/routes
* k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2

controller-runtime was also upgraded to match to Kube 1.27.

Added some new methods for the FakeClientWrapper to satisfy the client
interface.

Added arguments for CCM webhooks introduced in the k8s.io/cloud-provider
library.

Signed-off-by: Nolan Brubaker <nolan@nbrubaker.com>

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

I wasn't 100% sure if the webhook arguments are correct. As far as I could tell, they're not used in this CCM, but since this CCM uses k8s.io/cloud-provider directly, the arguments were necessary.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by lubronzhan** [comment]
Thanks for the PR!
/ok-to-test

**Comment by lubronzhan** [comment]
Hi could you fix the test?

**Comment by lubronzhan** [comment]
nvm, just see you are already fixing it. Thanks

**Comment by nrb** [comment]
Yep, looking at it now.

**Comment by nrb** [comment]
Discussed in [Slack](https://kubernetes.slack.com/archives/C9PGCDKV5/p1685133527950339?thread_ts=1685118713.324199&cid=C9PGCDKV5) - the presubmit tests are working locally for me, but Prow has a [Go 1.19/Kube 1.25](https://github.com/kubernetes/test-infra/blob/bc91fd978b95a6f0cb91196577751a169b444d11/config/jobs/kubernetes/cloud-provider-vsphere/cloud-provider-vsphere-config.yaml) image for this repo.

**Comment by lubronzhan** [comment]
Ok submited a PR to fix it https://github.com/kubernetes/test-infra/pull/29614

**Comment by lubronzhan** [comment]
The PR was merged
/retest

**Comment by lubronzhan** [comment]
I think the go.mod in e2e test also needs to be updated. The e2e test is failing to compile @nrb 
https://github.com/kubernetes/cloud-provider-vsphere/blob/master/test/e2e/go.mod

File: test/e2e/go.mod [hunk 1]
```diff
@@ -1,16 +1,16 @@
 module k8s.io/cloud-provider-vsphere/test/e2e
 
-go 1.19
+go 1.20
```
Comment (nrb) [inline test/e2e/go.mod:3]: @lubronzhan The e2e dependencies are updated here, but I'll check out why there's errors about the Informer types. I know some signatures have changed when bumping other projects to 1.27.

**Comment by lubronzhan** [comment]
Hm CAPI in test go.mod might need to be updated to 1.4.1

**Comment by nrb** [comment]
After some experimentation, I think CAPI itself needs to upgrade to controller-runtime v0.15.0, but currently that project is having issues with Kubernetes 1.27.

I expect this latest push, which bumps CAPI to 1.4.2 and controller-runtime to v0.15.0, will fail with this:

```shell
❯ go test e2e_suite_test.go
# sigs.k8s.io/cluster-api/api/v1beta1
../../../../go/1.20.1/pkg/mod/sigs.k8s.io/cluster-api@v1.4.2/api/v1beta1/machine_webhook.go:45:27: cannot use &Machine{} (value of type *Machine) as admission.Validator value in variable declaration: *Machine does not implement admission.Validator (wrong type for method ValidateCreate)
                have ValidateCreate() error
                want ValidateCreate() (admission.Warnings, error)
../../../../go/1.20.1/pkg/mod/sigs.k8s.io/cluster-api@v1.4.2/api/v1beta1/machinedeployment_webhook.go:56:27: cannot use &MachineDeployment{} (value of type *MachineDeployment) as admission.Validator value in variable declaration: *MachineDeployment does not implement admission.Validator (wrong type for method ValidateCreate)
                have ValidateCreate() error
                want ValidateCreate() (admission.Warnings, error)
../../../../go/1.20.1/pkg/mod/sigs.k8s.io/cluster-api@v1.4.2/api/v1beta1/machinedeployment_webhook.go:61:16: assignment mismatch: 2 variables but admission.NewDecoder returns 1 value
../../../../go/1.20.1/pkg/mod/sigs.k8s.io/cluster-api@v1.4.2/api/v1beta1/machinehealthcheck_webhook.go:63:27: cannot use &MachineHealthCheck{} (value of type *MachineHealthCheck) as admission.Validator value in variable declaration: *MachineHealthCheck does not implement admission.Validator (wrong type for method ValidateCreate)
                have ValidateCreate() error
                want ValidateCreate() (admission.Warnings, error)
../../../../go/1.20.1/pkg/mod/sigs.k8s.io/cluster-api@v1.4.2/api/v1beta1/machineset_webhook.go:45:27: cannot use &MachineSet{} (value of type *MachineSet) as admission.Validator value in variable declaration: *MachineSet does not implement admission.Validator (wrong type for method ValidateCreate)
                have ValidateCreate() error
                want ValidateCreate() (admission.Warnings, error)
FAIL    command-line-arguments [build failed]
FAIL
```

Checking [the ValidateCreate](https://github.com/kubernetes-sigs/cluster-api/blob/v1.4.2/api/v1beta1/machineset_webhook.go#L80) method on CAPI, it does indeed have the old signature. There's a [commit](https://github.com/kubernetes-sigs/cluster-api/commit/ea0dc4da091b13bea47dbb6d41340c1f434bac5d) on main currently that fixes it, but it appears to be unreleased.

---

CAPI 1.4.2 [depends on controller-runtime v0.14.5](https://github.com/kubernetes-sigs/cluster-api/blob/v1.4.2/go.mod#L45) currently. v0.14.6 doesn't appear to introduce major changes, so I modified the `go.mod` to try that. I get a failure with a different signature mismatch.
`
```shell
❯ go test e2e_suite_test.go
# sigs.k8s.io/controller-runtime/pkg/cache
../../../../go/1.20.1/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/cache/multi_namespace_cache.go:308:9: cannot use handles (variable of type map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration) as "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration value in return statement: map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration does not implement "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration (missing method HasSynced)
../../../../go/1.20.1/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/cache/multi_namespace_cache.go:321:9: cannot use handles (variable of type map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration) as "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration value in return statement: map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration does not implement "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration (missing method HasSynced)
../../../../go/1.20.1/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/cache/multi_namespace_cache.go:326:17: impossible type assertion: h.(map[string]toolscache.ResourceEventHandlerRegistration)
        map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration does not implement "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration (missing method HasSynced)
FAIL    command-line-arguments [build failed]
FAIL
```

This problem was [fixed already](https://github.com/kubernetes-sigs/controller-runtime/commit/13bba740ff6ea05a09b91e7b625fbcbdfea93d2f), and is included in v0.15.0.

---

I think this PR might be on hold until CAPI can upgrade to Kube 1.27 and controller-runtime v0.15.0. There's already at least one [issue](https://github.com/kubernetes-sigs/cluster-api/issues/8764) regarding the Kubernetes 1.27 support.

File: test/e2e/go.mod (hunk 1 above)
Comment (nrb) [reply test/e2e/go.mod:3]: Errors are due to CAPI/controller-runtime version incompatibilities.

**Comment by killianmuldoon** [comment]
Chiming in because I saw the issue referenced :slightly_smiling_face: 

>I think this PR might be on hold until CAPI can upgrade to Kube 1.27 and controller-runtime v0.15.0. 

CAPI won't update controller runtime on the v1.4.X series. The first release to support controller runtime v0.15.0 will be CAPI v1.5.0 which is planned for release on July 25th. 

>There's already at least one https://github.com/kubernetes-sigs/cluster-api/issues/8764 regarding the Kubernetes 1.27 support. 

There's no issues with Kubernetes 1.27 support in CAPI AFAIK. The main branch has already been updated to K8s v1.27 libraries and there's periodic tests for 1.27 clusters (as well as alpha 1.28 clusters). The issue linked above is a mismatch in  test-infrastructure images.

**Comment by nrb** [comment]
@lubronzhan Given that there won't be a CAPI release with the compatibility code until July, what are your thoughts on pinning the e2e's go.mod to a commit or pre-release of CAPI?

I figured I'd ask before spending time on trying to get that to work if you'd just like to wait.

**Comment by lubronzhan** [comment]
Thanks @killianmuldoon and @nrb 
I think it's OK to pin to a pre-release. We can bump that later. Don't want to block your PR

**Comment by nrb** [comment]
Ok, I'm focusing on some other work today but I'll work on updating this tomorrow.

**Comment by lubronzhan** [comment]
Hi @nrb any updates? Or you can give me the access to the branch I can work on it

**Comment by nrb** [comment]
@lubronzhan Sorry about that delay; I've pushed an update that compiles locally.

**Comment by lubronzhan** [comment]
/lgtm
/approve
Thanks!

**Comment by lubronzhan** [comment]
https://github.com/kubernetes/cloud-provider/commit/ad15f1c5238c13062d68e042bbd98aebe6e6ea85
I think the webhook framework parameters are good. It's just a new feature. I didn't find PR associated with it. I guess it's just a new feature added for us to add webhook

**Comment by nrb** [comment]
Yeah, it looks like that's exposing options should you choose to use them. For the existing code it should be inert.

**Pull request merged**